### DISCUSSION TITLE
Directus extensions types documentation

### DISCUSSION THREAD

**Discussion by Enhed** [directus/directus#12563] [Archive]
how to get docs for types in directus with using `@directus/extensions-sdk`?  

![image](https://user-images.githubusercontent.com/36981278/161543914-ff8259d5-fdec-4b98-8cba-3238f2868a45.png)

**Comment by rijkvanzanten** [comment]
Looks like the typing for the api extensions context doesn't currently type services/extensions explicitly:

https://github.com/directus/directus/blob/4cb8e8f0ca3fc479c443adfc8c1a04d444f4ae87/packages/shared/src/types/extensions.ts#L86-L87

**Reply by samdze** [reply]
It would be very helpful to have complete types!

**Reply by abdonrd** [reply]
I was going to report this problem, but I see that it was already reported!

<img width="539" alt="Screen Shot 2022-07-12 at 10 36 56" src="https://user-images.githubusercontent.com/1007051/178448677-a611527b-f2a7-446b-857f-9d3076a58d69.png">

**Reply by abdonrd** [reply]
We have one JS extension with types like this:

<img width="603" alt="Screen Shot 2022-07-12 at 11 24 28" src="https://user-images.githubusercontent.com/1007051/178457680-637528d2-97e6-417d-947f-83359ade38cc.png">

But how we can fix the core types of the `packages/shared/src/types/extensions.ts`?
I'm not sure about how the mono-repo works.

Because probably this is wrong:
```typescript
import * as Services from '../../../../api/src/services';
// ...
services: Services;
```

**Reply by rijkvanzanten** [reply]
That last example _should_ be fine, as the `@directus/shared` package goes through a build step before being distributed, though @nickrum could enlighten us more on that

**Reply by LudovicSterlin** [reply]
Would it be OK to open an issue on adding the type of all the services (`ItemsService`, `FieldsService`, `RolesService`, ...) available in the `services` property of `ApiExtensionContext`. Would be really useful when coding extension in typescript. 
Plus, all the services seems to be already documented in `'directus/dist/services'`. But in the @directus/shared/ nothing is present.

It gets pretty annoying since the only way to access them is from the `services` parameter coming in the `HookExtensionContext` which is typed as any...

**Reply by azrikahar** [reply]
@LudovicSterlin we are tracking this in directus/docs#20 👍 Totally understand how difficult it can be. There was also a PR for it, but I believe it was temporarily closed due to some issues when it was still in this monorepo (conflicting vue version and vuepress etc). Good news is we have since moved docs into a separate repo, and the migration to VitePress was completed few days ago, so those are definitely steps in the right direction for this. Hope that's understandable.

**Reply by LudovicSterlin** [reply]
Definitely understandable, happy to learn about that being added in the doc.

However, for me, the issue is more about receiving the `services: any` in the hooks.

> Looks like the typing for the api extensions context doesn't currently type services/extensions explicitly:
https://github.com/directus/directus/blob/1fedc5bdfd7ef7f731ce0bddd6b2e1c50d69f120/packages/shared/src/types/extensions.ts#L132-L134

Is there a way to make a link between the `services` and all the `<service_name>.d.ts` available in `'directus/dist/services'` from `directus` package. Because those `.d.ts`file are really useful when your IDE find them you can code pretty easily, for now I import the _XxxxServices_ from those file to code and then remove the import to use the one from `services` parameter when testing.

Maybe something like this:
(I'm not really confident with the Type thing since i don't know how is really constructed the services object)
(Also the import from `'../../../../api/src/services'` seems to be the difficult thing to do since for now there is only shared types, and now we need to share classes declaration from the api)
```ts
import { ActivityService, AssetsService, AuthenticationService, AuthorizationService, CollectionsService, DashboardsService, ExportService, FieldsService, FilesService, FlowsService, FoldersService, GraphQLService, ImportService, ItemsService, MailService, MetaService, NotificationsService, OperationsService, PanelsService, PayloadService, PermissionsService, PresetsService, RelationsService, RevisionsService, RolesService, ServerService, SettingsService, SharesService, SpecificationService, TFAService, UsersService, UtilsService, WebhooksService } from '../../../../api/src/services';

export interface Type<T> extends Function { new(...args: any[]): T; }

export type Services = {
	ActivityService: Type<ActivityService>,
	AssetsService: Type<AssetsService>,
	AuthenticationService: Type<AuthenticationService>,
	AuthorizationService: Type<AuthorizationService>,
	CollectionsService: Type<CollectionsService>,
	DashboardsService: Type<DashboardsService>,
	ExportService: Type<ExportService>,
	FieldsService: Type<FieldsService>,
	FilesService: Type<FilesService>,
	FlowsService: Type<FlowsService>,
	FoldersService: Type<FoldersService>,
	GraphQLService: Type<GraphQLService>,
	ImportService: Type<ImportService>,
	ItemsService: Type<ItemsService>,
	MailService: Type<MailService>,
	MetaService: Type<MetaService>,
	NotificationsService: Type<NotificationsService>,
	OperationsService: Type<OperationsService>,
	PanelsService: Type<PanelsService>,
	PayloadService: Type<PayloadService>,
	PermissionsService: Type<PermissionsService>,
	PresetsService: Type<PresetsService>,
	RelationsService: Type<RelationsService>,
	RevisionsService: Type<RevisionsService>,
	RolesService: Type<RolesService>,
	ServerService: Type<ServerService>,
	SettingsService: Type<SettingsService>,
	SharesService: Type<SharesService>,
	SpecificationService: Type<SpecificationService>,
	TFAService: Type<TFAService>,
	UsersService: Type<UsersService>,
	UtilsService: Type<UtilsService>,
	WebhooksService: Type<WebhooksService>,
};
```

and then 

```ts
export type ApiExtensionContext = { 
 	services: Services; 
 	exceptions: any; 
	[...]
};
```

Would love to participate in a PR with some hints on how to make the bridge.

**Reply by azrikahar** [reply]
I think that does sound similar to what was mentioned here: https://github.com/directus/directus/discussions/3025#discussioncomment-2006511. @nickrum any thoughts on this?

**Reply by LudovicSterlin** [reply]
Looking at #3025 and #11179 it looks similar.
However the issue raised at the end of the discussion by @johnhuffsmith https://github.com/directus/directus/discussions/3025#discussioncomment-2009398 is still present.

The difference I see is that we would have to move not only files from [/api/src/types](https://github.com/directus/directus/tree/main/api/src/types) to [/packages/shared/src/types](https://github.com/directus/directus/tree/main/packages/shared/src/types), but also from [/api/src/services](https://github.com/directus/directus/tree/main/api/src/services).

One first solution could be to use something like `BaseException` https://github.com/directus/directus/blob/ef939072cc0b41575f9f0507e5798ea2d645640b/packages/shared/src/exceptions/base.ts#L1-L3
_For `services` there is https://github.com/directus/directus/blob/ef939072cc0b41575f9f0507e5798ea2d645640b/api/src/types/services.ts#L11-L13 but not all services implement it (only the one that `extends ItemsService`)._

Which would allow making a ServiceOverview/ExceptionsOverview type
```ts
export type ServiceOverview = {
	ActivityService: BaseService,
	AssetsService: BaseService,
        // [...]
	UtilsService: BaseService,
	WebhooksService: BaseService,
};
```
Benefits:
+ Allow to specify the structure of `services` and `exceptions` in `ApiExtensionContext`
+ Should be easier to implement because this remove the difficulty of the big import of all the services/exceptions inside `/packages/shared/src/types`
    ```
    import { ActivityService, AssetsService, AuthenticationService, AuthorizationService, CollectionsService, DashboardsService, ExportService, FieldsService, FilesService, FlowsService, FoldersService, GraphQLService, ImportService, ItemsService, MailService, MetaService, NotificationsService, OperationsService, PanelsService, PayloadService, PermissionsService, PresetsService, RelationsService, RevisionsService, RolesService, ServerService, SettingsService, SharesService, SpecificationService, TFAService, UsersService, UtilsService, WebhooksService } from '../../../../api/src/services';
    ```
Missing features:
- Giving all the specific function available for each service
- Specifying the right way (the arguments expected by the constructor) to create each specific exceptions.

**Reply by nickrum** [reply]
Importing the services through `import * as Services from '../../../../api/src/services'` inside the `@directus/shared` package wouldn't work because currently this package is built using the typescript compiler, which means imports will be kept as is. Also, directly importing files from another package might not be a good idea to begin with.

@LudovicSterlin's idea sounds like great temporary solution to at least improve the situation, but we should still focus on finding a permanent solution without any drawbacks for extension creators. The best we can do is probaly still to create dedicated interfaces for each service inside shared and let the actual service classes implement those.

Another solution might be to move the `define*()` functions for API extensions to the API package and to let extensions import them from there.

**Comment by LudovicSterlin** [comment]
> Importing the services through `import * as Services from '../../../../api/src/services'` inside the `@directus/shared` package wouldn't work because currently this package is built using the typescript compiler, which means imports will be kept as is. Also, directly importing files from another package might not be a good idea to begin with.
> 
> @LudovicSterlin's idea sounds like great temporary solution to at least improve the situation, but we should still focus on finding a permanent solution without any drawbacks for extension creators. The best we can do is probaly still to create dedicated interfaces for each service inside shared and let the actual service classes implement those.
> 
> Another solution might be to move the `define*()` functions for API extensions to the API package and to let extensions import them from there.

@azrikahar  @nickrum I continue our talk in the main thread of this discussion.

I started implementing something for a future PR [here](https://github.com/LudovicSterlin/directus/tree/feat-add-services-interface-in-shared). If this look on the good way, we could maybe open an issue? So, I can submit a PR for this ?

**Reply by LudovicSterlin** [reply]
I opened #15711 to try to go forward with this. Could we transform this discussion into an issue, so the PR is attached to an Issue ?

**Reply by nickrum** [reply]
As this is not really an issue, I think we should keep it a discussion. Your PR doesn't strictly need to be attached to an issue if it is an improvement and not fixing any issues.

**Reply by rijkvanzanten** [reply]
> As this is not really an issue, I think we should keep it a discussion. Your PR doesn't strictly need to be attached to an issue if it is an improvement and not fixing any issues.

We do as per 5 hours ago 😁  https://github.com/directus/directus/pull/16246/files

Reviewing and merging PRs that weren't previously discussed and accepted is becoming too much of a burden on the core maintainer team 👍🏻

**Reply by LudovicSterlin** [reply]
So, do I need to do anything on my side to better document the discussion and what I tried to implement ? Or is it just a question of human availability and I should just wait that someone has time to review my PR ?

**Reply by rijkvanzanten** [reply]
It's a question of human availability. This is not a tweak we were planning on making / maintaining right now, and the PR is over a thousand lines of code changed that doesn't have any existing test coverage. Getting that reviewed and merged is a multi-hour process, which is time we can't take away from the other in progress tasks the core team is working on. We have to prioritize based on impact on the community, and unfortunately this post has only gotten a handful of upvotes.

**Comment by TyrionFront** [comment]
@rijkvanzanten , is theere any chance that you at least give back the types for extensions functions that were previously ? Just like in version 9.20.4 for example:

<img width="805" alt="Screenshot 2022-11-23 at 22 00 33" src="https://user-images.githubusercontent.com/18406903/203636532-93a6d1be-2b9a-4aa6-84ff-907c998ccd0f.png">
<img width="556" alt="Screenshot 2022-11-23 at 22 00 40" src="https://user-images.githubusercontent.com/18406903/203636535-f6326c43-e9a4-43e8-999b-578c3bc1dc2a.png">

because in current state it looks more like regression and not like an improvement

<img width="586" alt="Screenshot 2022-11-23 at 22 04 58" src="https://user-images.githubusercontent.com/18406903/203636814-7bbe0920-d632-4f42-80d0-a0b38a853c83.png">
